databuilder/extractor/minio_stats_extractor.py

```diff
@@ -0,0 +1,113 @@
+from pyhocon import ConfigTree
+from typing import Any, Iterator, List
+import boto3
+from databuilder.extractor.base_extractor import Extractor
+from databuilder.utils.minio_spec import MinioSpecUtils
+from databuilder.models.table_stats import TableColumnStats
+#from databuilder.utils.spark_driver_dev import initSparkSessionLocal
+from databuilder.utils.pyspark_stats import get_stats_by_column
+from pyspark.sql.session import SparkSession
+
+
+class MinioStatsExtractor(Extractor):
+
+    """
+    An Extractor that extracts meta data from minio and stores them in amundsen.
+    """
+    # CONFIG KEYS
+    ACCESS_KEY = 'access_key'
+    SECRET_KEY = 'secret_key'
+    BUCKET_NAME = 'bucket_name'
+    ENDPOINT_URL = 'endpoint_url'
+    SPARK_SESSION_KEY = 'spark_session'
+
+    def init(self, conf: ConfigTree) -> None:
+        """
+        :param conf:
+        """
+        self.conf = conf
+        self._extract_iter = None
+
+        self.bucket_name = conf.get_string(MinioStatsExtractor.BUCKET_NAME)
+        self.endpoint_url = conf.get_string(MinioStatsExtractor.ENDPOINT_URL)
+        self.access_key = conf.get_string(MinioStatsExtractor.ACCESS_KEY)
+        self.secret_key = conf.get_string(MinioStatsExtractor.SECRET_KEY)
+        self.spark_session: SparkSession = conf.get(MinioStatsExtractor.SPARK_SESSION_KEY)
+        #self.spark_session = initSparkSessionLocal()
+
+        self.db = 'minio'
+        self.schema = 'minio'
+
+        self.s3 = boto3.resource('s3',
+                                 endpoint_url=self.endpoint_url,
+                                 aws_access_key_id=self.access_key,
+                                 aws_secret_access_key=self.secret_key,
+                                 region_name='us-east-1')
+        # TODO: use spark session from the arg parser using Spark/Kubernetes deployment
+
+    def get_stats_iter(self) -> Iterator[dict]:
+        """Get a stats iter of column stats dictionaries."""
+        stat_objs = []
+        keys = self.get_dataset_keys()
+
+        for key in keys:
+            stat_objs.extend(self.get_file_stats(key))
+
+        print("stat_objs: ", stat_objs)
+        return iter(stat_objs)
+
+    def get_dataset_keys(self) -> List[str]:
+        """Get a list of keys in an S3 bucket."""
+        dataset_paths = set()
+        result = self.s3.meta.client.list_objects(Bucket=self.bucket_name, Delimiter='/',
+                                                  Prefix='v0/')
+        for o in result.get('CommonPrefixes', []):
+            path = o.get('Prefix')
+            if MinioSpecUtils.path_to_dataset_name(path) is None:
+                continue
+            dataset_paths.add(path)
+        return dataset_paths
```

Comment:
There's a lot of duplicated code w/ minio_extractor that I assume will be deduplicated in the final version of this PR.

Reply:
yeah we should work on dedeuplicating, though I will note this function is slightly different than the other because it doesn't return an iterator. The only iterator in this new class is for the stats objects. (I just also noticed the return type is incorrect which I'll fix as well).